File: vox/swift/vox-runtime/Sources/VoxRuntime/Conduit.swift
import Foundation

public protocol Conduit: Sendable {
    func send(_ message: Message) async throws
    func recv() async throws -> Message?
    func setMaxFrameSize(_ size: Int) async throws
    func close() async throws
}

/// A conduit over a raw `Link`: the envelope rides phon (`encodeMessage` /
/// `decodeMessage`). The decoder uses the peer's advertised Message schema
/// (from the handshake) against the local reader.
/// r[impl conduit]
/// r[impl conduit.bare]
public final class BareConduit: Conduit, @unchecked Sendable {
    public let link: any Link
    private let decode: MessageDecoder

    public init(link: any Link, peerMessageSchema: [UInt8]) {
        self.link = link
        self.decode = buildMessageDecoder(peerMessageSchema: peerMessageSchema)
    }

    public func send(_ message: Message) async throws {
        try await link.sendFrame(encodeMessage(message))
    }

    public func recv() async throws -> Message? {
        guard let bytes = try await link.recvFrame() else { return nil }
        return try decode(bytes)
    }

    public func setMaxFrameSize(_ size: Int) async throws {
        try await link.setMaxFrameSize(size)
    }

    public func close() async throws {
        try await link.close()
    }
}

extension Link {
    public func bareConduit(peerMessageSchema: [UInt8]) -> BareConduit {
        BareConduit(link: self, peerMessageSchema: peerMessageSchema)
    }
}

File: vox/typescript/packages/vox-core/src/conduit.ts
import {
  buildMessageDecoder,
  decodeMessageWith,
  encodeMessage,
  type Message,
  type MessageDecoder,
} from "@bearcove/vox-wire";
import type { Link } from "./link.ts";
import { voxLogger } from "./logger.ts";

export interface Conduit<T> {
  send(item: T): Promise<void>;
  recv(): Promise<T | null>;
  close(): void;
  isClosed(): boolean;
}

/**
 * Build the envelope decoder from the peer's `Message` schema (exchanged
 * in the handshake, as phon schema-closure bytes) against ours.
 * With no peer schema it degenerates to writer==reader.
 */
// r[impl conduit.typeplan]
export function buildMessageDecodePlan(peerSchemaBytes?: Uint8Array): MessageDecoder {
  return buildMessageDecoder(peerSchemaBytes);
}

// r[impl conduit]
// r[impl conduit.bare]
export class BareConduit implements Conduit<Message> {
  private readonly link: Link;
  private readonly decoder: MessageDecoder;

  constructor(link: Link, decoder: MessageDecoder | null = null) {
    this.link = link;
    this.decoder = decoder ?? buildMessageDecoder();
  }

  async send(item: Message): Promise<void> {
    await this.link.send(encodeMessage(item));
  }

  async recv(): Promise<Message | null> {
    const payload = await this.link.recv();
    if (!payload) {
      return null;
    }
    try {
      return decodeMessageWith(this.decoder, payload);
    } catch (e) {
      voxLogger()?.error(`[vox:conduit] decode failed (${payload.length} bytes):`, e);
      throw e;
    }
  }

  close(): void {
    this.link.close();
  }

  isClosed(): boolean {
    return this.link.isClosed();
  }
}
